core/store/src/db/splitdb.rs

```diff
@@ -0,0 +1,236 @@
+use crate::DBCol;
+
+use super::StoreStatistics;
+
+/// A database that provides access to the hot and cold databases.
+///
+/// For hot-only columns it always reads from the hot database only. For cold
+/// columns it reads from hot first and if the value is present it returns it.
+/// If the value is not present it reads from the cold database.
+///
+/// The iter and iter_prefix methods return either the iterator of the hot
+/// database or of the cold database depending on if the column is hot or cold.
+///
+/// The iter_raw_bytes method is not supported but it falls back to returning
+/// the hot storage iterator.
+///
+/// This is a read-only database that doesn't support writing. This is because
+/// it would be not clear what database should be written to.
+pub struct SplitDB {
+    hot: Arc<dyn Database>,
+    cold: Arc<dyn Database>,
+}
+
+impl SplitDB {
+    pub fn new(hot: Arc<dyn Database>, cold: Arc<dyn Database>) -> Arc<Self> {
+        return Arc::new(SplitDB { hot, cold });
+    }
+}
+
+impl Database for SplitDB {
+    /// Returns raw bytes for given `key` ignoring any reference count decoding
+    /// if any.
+    ///
+    /// First tries to read the data from the hot db and returns it if found.
+    /// Then it tries to read the data from the cold db and returns the result.
+    fn get_raw_bytes(&self, col: DBCol, key: &[u8]) -> io::Result<Option<DBSlice<'_>>> {
+        if let Some(hot_result) = self.hot.get_raw_bytes(col, key)? {
+            return Ok(Some(hot_result));
+        }
+        if col.is_cold() {
+            return self.cold.get_raw_bytes(col, key);
+        }
+        Ok(None)
+    }
+
+    /// Returns value for given `key` forcing a reference count decoding.
+    ///
+    /// **Panics** if the column is not reference counted.
+    ///
+    /// First tries to read the data from the hot db and returns it if found.
+    /// Then it tries to read the data from the cold db and returns the result.
+    fn get_with_rc_stripped(&self, col: DBCol, key: &[u8]) -> io::Result<Option<DBSlice<'_>>> {
+        assert!(col.is_rc());
+
+        if let Some(hot_result) = self.hot.get_with_rc_stripped(col, key)? {
+            return Ok(Some(hot_result));
+        }
+        if col.is_cold() {
+            return self.cold.get_with_rc_stripped(col, key);
+        }
+        Ok(None)
+    }
+
+    /// Iterate over all items in given column in lexicographical order sorted
+    /// by the key.
+    ///
+    /// For cold columns returns the cold iterator.
+    /// For other columns returns the hot iterator.
+    ///
+    /// Keep in mind that the cold db is typically behind by a few blocks. It
+    /// may be behind by up to gc_num_epochs_to_keep epochs under acceptable
+    /// conditions or more if the cold store loop is falling behind under
+    /// erroneous conditions.
+    fn iter<'a>(&'a self, column: DBCol) -> DBIterator<'a> {
+        match column.is_cold() {
+            false => self.hot.iter(column),
```

Comment:
The comment states that cold db is typically behind.
Is it acceptable to iterate only over cold db? Should the iteration be done over both cold and hot?

Reply:
Great question but I don't have the answer. In this PR I wanted to get a minimal sensible implementation so I did it this way. For hot only columns it is correct to only iterate over hot. For cold columns we may want iterate over both but then there is the issue of duplicated entries. To make things complicated when copying data to cold, for some columns, we change the format of the key. IIRC for some column we remove shard id from the key and for some other we change endianness (don't ask me why). This would change the order of iteration and make duplicate removal hard. 

TLDR: let's keep it as is for now and iterate (pun intended) on this implementation as needed.